platform/o.n.bootstrap/src/org/netbeans/ProxyClassLoader.java

```diff
@@ -468,34 +468,35 @@ protected Package getPackage(String name) {
      * @return located package, or null
      */
     protected Package getPackageFast(String name, boolean recurse) {
-        synchronized (packages) {
-            Package pkg = packages.get(name);
-            if (pkg != null) {
-                return pkg;
-            }
-            if (!recurse) {
-                return null;
-            }
-            String path = name.replace('.', '/');
-            for (ProxyClassLoader par : this.parents.loaders()) {
-                if (!shouldDelegateResource(path, par))
-                    continue;
-                pkg = par.getPackageFast(name, false);
-                if (pkg != null) break;
-            }
-            // pretend the resource ends with "/". This works better with hidden package and
-            // prefix-based checks.
-            if (pkg == null && shouldDelegateResource(path + "/", null)) {
-                // Cannot access either Package.getSystemPackages nor ClassLoader.getPackage
-                // from here, so do the best we can though it will cause unnecessary
-                // duplication of the package cache (PCL.packages vs. CL.packages):
-                pkg = super.getPackage(name);
+        Package pkg = packages.get(name);
+        if (pkg != null) {
+            return pkg;
+        }
+        if (!recurse) {
+            return null;
+        }
+        String path = name.replace('.', '/');
+        for (ProxyClassLoader par : this.parents.loaders()) {
+            if (!shouldDelegateResource(path, par)) {
+                continue;
             }
+            pkg = par.getPackageFast(name, false);
             if (pkg != null) {
-                packages.put(name, pkg);
+                break;
             }
-            return pkg;
         }
+        // pretend the resource ends with "/". This works better with hidden package and
+        // prefix-based checks.
+        if (pkg == null && shouldDelegateResource(path + "/", null)) {
+            // Cannot access either Package.getSystemPackages nor ClassLoader.getPackage
+            // from here, so do the best we can though it will cause unnecessary
+            // duplication of the package cache (PCL.packages vs. CL.packages):
+                pkg = super.getPackage(name);
+        }
+        if (pkg != null) {
+            packages.put(name, pkg);
```

Comment:
You are correct, there can be a replaced entry in the map between the line 471 and line 497. I will add again the `synchronized(packages)` block to prevent any nondeterministic result.